### PR TITLE
fix bugs for franka robot control

### PR DESCRIPTION
**Pull request by Zhouxunzhe**
# [Bug fix] fix some bugs for franka robot control

fix some bugs for franka robot control to run `quickstart`.

### CODE REVIEW

File: robosuite/environments/base.py [hunk 1]
```diff
@@ -108,7 +108,8 @@ def __init__(
         # Rendering-specific attributes
         self.has_renderer = has_renderer
         # offscreen renderer needed for on-screen rendering
-        self.has_offscreen_renderer = (has_renderer and renderer != "mjviewer") or has_offscreen_renderer
+        self.has_offscreen_renderer = has_renderer or has_offscreen_renderer
+        # self.has_offscreen_renderer = (has_renderer and renderer != "mjviewer") or has_offscreen_renderer
```
Comment (zhuyifengzju) [inline robosuite/environments/base.py:112]: remove comemnts

File: robosuite/environments/base.py [hunk 2]
```diff
@@ -182,18 +190,20 @@ def initialize_renderer(self):
...
-            self.viewer = MjviewerRenderer(env=self, camera_id=camera_id)
+        # elif self.renderer == "mjviewer":
+        #     from robosuite.renderers.mjviewer.mjviewer_renderer import MjviewerRenderer
+        #
+        #     if self.render_camera is not None:
+        #         camera_id = self.sim.model.camera_name2id(self.render_camera)
+        #     else:
+        #         camera_id = None
+        #     self.viewer = MjviewerRenderer(env=self, camera_id=camera_id, sim=self.sim)
+        #     self.viewer.set_camera(camera_id)
```
Comment (zhuyifengzju) [inline robosuite/environments/base.py:206]: Please remove all comments like this

File: robosuite/renderers/mjviewer/mjviewer_renderer.py [hunk 3]
```diff
@@ -1,15 +1,28 @@
...
 class MjviewerRenderer:
-    def __init__(self, env, camera_id=None, cam_config=None):
+    def __init__(self, env, camera_id=None, cam_config=None, sim=None):
         self.env = env
         self.camera_id = camera_id
         self.viewer = None
         self.camera_config = cam_config
 
+        self.width = 1280
+        self.height = 800
```
Comment (zhuyifengzju) [inline robosuite/renderers/mjviewer/mjviewer_renderer.py:14]: Will these values be flexible to change? We want some configuration more flexible than writing fixed values like this.

File: robosuite/robots/fixed_base_robot.py [hunk 4]
```diff
@@ -219,11 +219,11 @@ def control(self, action, policy_step=False):
...
                 # Update arm-specific proprioceptive values
                 self.recent_ee_forcetorques[arm].push(np.concatenate((self.ee_force[arm], self.ee_torque[arm])))
-                self.recent_ee_pose[arm].push(np.concatenate((controller.ee_pos, T.mat2quat(controller.ee_ori_mat))))
-                self.recent_ee_vel[arm].push(np.concatenate((controller.ee_pos_vel, controller.ee_ori_vel)))
+                self.recent_ee_pose[arm].push(np.concatenate((controller.ref_pos, T.mat2quat(controller.ref_ori_mat))))
+                self.recent_ee_vel[arm].push(np.concatenate((controller.ref_pos_vel, controller.ref_ori_vel)))
 
                 # Estimation of eef acceleration (averaged derivative of recent velocities)
-                self.recent_ee_vel_buffer[arm].push(np.concatenate((controller.ee_pos_vel, controller.ee_ori_vel)))
+                self.recent_ee_vel_buffer[arm].push(np.concatenate((controller.ref_pos_vel, controller.ref_ori_vel)))
```
Comment (zhuyifengzju) [inline robosuite/robots/fixed_base_robot.py:226]: We want to use the ee_pos, ee_pos_vel defined by the controller to make implementation more intuitive. Can you reverse back these changes?

File: robosuite/robots/robot.py [hunk 5]
```diff
@@ -536,7 +536,8 @@ def dof(self):
             int: degrees of freedom of the robot (with grippers).
         """
         # Get the dof of the base robot model
-        dof = super().dof
+        dof = self.robot_model.dof
+        # dof = super().dof
```
Comment (zhuyifengzju) [inline robosuite/robots/robot.py:540]: remove comment.

File: robosuite/utils/opencv_renderer.py [hunk 6]
```diff
@@ -46,5 +76,9 @@ def close(self):
         # NOTE: assume that @sim will get cleaned up outside the renderer - just delete the reference
         self.sim = None
 
+        if self.viewer is not None:
+            self.viewer.close()
+            self.viewer = None
```
Comment (zhuyifengzju) [inline robosuite/utils/opencv_renderer.py:81]: I didn't quite get this part. The implementation looks like mjviewer, but isn't this class designed for OpenCV? Can you double check or explain here what's reason changing this file?

File: robosuite/environments/base.py [hunk 7]
```diff
@@ -169,11 +170,18 @@ def __init__(
...
+            self.renderer_config['sim'] = self.sim
 
         if self.renderer == "mujoco" or self.renderer == "default":
-            pass
+            # Set the camera angle for viewing
+            camera_id = self.sim.model.camera_name2id(self.render_camera) if self.render_camera is not None else 0
+
+            # TODO(zxz): modify here to set renderer args
+            self.viewer = OpenCVRenderer(env=self, camera_id=camera_id, **self.renderer_config)
+            self.viewer.set_camera(camera_id)
```
Comment (zhuyifengzju) [inline robosuite/environments/base.py:184]: See the comment for opencv_renderer.py

**Review by zhuyifengzju — CHANGES_REQUESTED** [review]
Thanks for creating the PR. There are some details that need to be fixed and then tested thoroughly before we merge this PR.

File: robosuite/renderers/mjviewer/mjviewer_renderer.py (hunk 3 above)
Comment (Zhouxunzhe) [reply robosuite/renderers/mjviewer/mjviewer_renderer.py:14]: I just follow the structure of OpenCVRenderer (https://github.com/ARISE-Initiative/robosuite/blob/6ec07a1ede31e1ede93c3d29fce4a7fefe0cad1f/robosuite/utils/opencv_renderer.py#L11). Maybe you could pass the configuration through some args like screen_config?

File: robosuite/robots/fixed_base_robot.py (hunk 4 above)
Comment (Zhouxunzhe) [reply robosuite/robots/fixed_base_robot.py:226]: No problem! Here I just want to quick-solve the issues then.

File: robosuite/utils/opencv_renderer.py (hunk 6 above)
Comment (Zhouxunzhe) [reply robosuite/utils/opencv_renderer.py:81]: Since I couldn't use mjviewer for rendering then, I decided to use OpenCV. However, for real-time rendering, we need to use the function `update` to update robot state, which uses `mujoco.viewer`. When I was trying to transfer this logic from mjviewer to OpenCV, I need to init the viewer and close the viewer. That's why you can see many implementations that look like mjviewer in OpenCV. I was just for convenience. You can modify these files according to your needs. Hope the explanation clear for you!

**Comment by Zhouxunzhe** [comment]
I have pushed a new version, please check it out.

File: robosuite/renderers/mjviewer/mjviewer_renderer.py [hunk 8]
```diff
@@ -1,15 +1,28 @@
...
+        self.sim = sim
+        self.camera_name = self.sim.model.camera_id2name(0)
+
     def render(self):
-        pass
+        # get frame with offscreen renderer (assumes that the renderer already exists)
```
Comment (snasiriany) [inline robosuite/renderers/mjviewer/mjviewer_renderer.py:19]: this render function is only supposed to be used for onscreen rendering. it's not supposed to return any values.

File: robosuite/models/robots/manipulators/panda_robot.py [hunk 9]
```diff
@@ -30,7 +30,7 @@ def default_gripper(self):
 
     @property
     def default_controller_config(self):
-        return {"right", "default_panda"}
+        return {"right": "default_panda"}
```
Comment (snasiriany) [inline robosuite/models/robots/manipulators/panda_robot.py:33]: fixed already in the latest code in `robocasa_v0.1`. thanks!

File: robosuite/environments/base.py [hunk 10]
```diff
@@ -516,7 +512,7 @@ def render(self):
         """
         Renders to an on-screen window.
         """
-        self.viewer.render()
+        return self.viewer.render()
```
Comment (snasiriany) [inline robosuite/environments/base.py:515]: this function is supposed to perform onscreen rendering rather than return any values

File: robosuite/robots/robot.py [hunk 11]
```diff
@@ -536,7 +536,7 @@ def dof(self):
             int: degrees of freedom of the robot (with grippers).
         """
         # Get the dof of the base robot model
-        dof = super().dof
+        dof = self.robot_model.dof
```
Comment (snasiriany) [inline robosuite/robots/robot.py:539]: noticed we have an extra `dof` function here that needed to be removed. thanks for finding this issue!

File: robosuite/utils/opencv_renderer.py (hunk 6 above)
Comment (snasiriany) [reply robosuite/utils/opencv_renderer.py:77]: I agree with @zhuyifengzju that the changes in this file should not be included

File: robosuite/renderers/mjviewer/mjviewer_renderer.py (hunk 3 above)
Comment (snasiriany) [reply robosuite/renderers/mjviewer/mjviewer_renderer.py:14]: controlling the width / height of the mjviewer is currently not supported, so this feature is not yet applicable here

File: robosuite/renderers/base.py [hunk 12]
```diff
@@ -22,6 +22,8 @@ def load_renderer_config(renderer):
         fname = "config/nvisii_config.json"
     elif renderer == "mjviewer":
         return {}
+    elif renderer == "mujoco":
```
Comment (snasiriany) [inline robosuite/renderers/base.py:25]: why this change?

File: robosuite/environments/base.py [hunk 13]
```diff
@@ -169,11 +169,16 @@ def __init__(
...
+        if self.renderer_config is None:
             self.renderer_config = load_renderer_config(self.renderer)
+            self.renderer_config['sim'] = self.sim
 
         if self.renderer == "mujoco" or self.renderer == "default":
-            pass
+            # Set the camera angle for viewing
+            camera_id = self.sim.model.camera_name2id(self.render_camera) if self.render_camera is not None else 0
+
+            self.viewer = OpenCVRenderer(env=self, camera_id=camera_id, **self.renderer_config)
```
Comment (snasiriany) [inline robosuite/environments/base.py:180]: skipping, as we will be sunseting OpenCVRenderer

**Review by snasiriany — COMMENTED** [review]
there were a few things here that we integrated into the codebase. thank you!
